### PR TITLE
kv: fix some bugs in memdb

### PR DESCRIPTION
**Pull request by tanghaodong25**
I'm trying to replace rocksdb with memdb to identify the overheads of rocksdb for bluestore, but I hit a few errors.
1. Can't kill ceph-osd process and then restart it.
2. if using memdb for bitmap allocator, it will hit assert fault in 'split_key' function.
3. fix iterator invalidation in memdb.

Signed-off-by: Haodong Tang <haodong.tang@intel.com>

### CODE REVIEW

File: src/kv/MemDB.cc [hunk 1]
```diff
@@ -425,6 +425,10 @@ bool MemDB::MDBWholeSpaceIteratorImpl::raw_key_is_prefixed(
     const string &prefix)
 {
   string p, k;
+  size_t pos = m_key_value.first.find(KEY_DELIM, 0);
+  if (pos == std::string::npos) {
+    return false;
+  }
```
Comment (liewegas) [inline src/kv/MemDB.cc]: hmm, why is there a key that doesn't have KEY_DELIM?

File: src/kv/MemDB.cc [hunk 2]
```diff
@@ -471,23 +476,25 @@ int MemDB::MDBWholeSpaceIteratorImpl:: prev()
...
   }
+  fill_current();
   return 0;
 }
 
 int MemDB::MDBWholeSpaceIteratorImpl::seek_to_last(const std::string &k)
 {
   std::lock_guard<std::mutex> l(*m_btree_lock_p);
-
+  if (!(*is_valid)) return -1;
```
Comment (liewegas) [inline src/kv/MemDB.cc]: fix whitespace please (statement on next line)

File: src/kv/MemDB.cc [hunk 3]
```diff
@@ -442,6 +445,7 @@ bufferlist MemDB::MDBWholeSpaceIteratorImpl::value()
 int MemDB::MDBWholeSpaceIteratorImpl::next()
 {
   std::lock_guard<std::mutex> l(*m_btree_lock_p);
+  if (!(*is_valid)) return -1;
```
Comment (liewegas) [inline src/kv/MemDB.cc]: here too

File: src/kv/MemDB.cc [hunk 4]
```diff
@@ -455,6 +459,7 @@ int MemDB::MDBWholeSpaceIteratorImpl::next()
 int MemDB::MDBWholeSpaceIteratorImpl:: prev()
 {
   std::lock_guard<std::mutex> l(*m_btree_lock_p);
+  if (!(*is_valid)) return -1;
```
Comment (liewegas) [inline src/kv/MemDB.cc]: and here

File: src/kv/MemDB.cc [hunk 5]
```diff
@@ -125,6 +125,7 @@ void MemDB::_load()
     dout(10) << __func__ << " Key:"<< key << dendl;
     m_btree[key] = datap;
   }
+  m_total_bytes += bytes_done;
```
Comment (liewegas) [inline src/kv/MemDB.cc]: put this in a different commit?

**Comment by liewegas** [comment]
rebase please?

File: src/kv/MemDB.cc (hunk 1 above)
Comment (tanghaodong25) [inline src/kv/MemDB.cc]: After checking for iterator invalidation, I no longer hit the 'assert_fail'. So I will delete this commit.

**Comment by tanghaodong25** [comment]
@liewegas have updated and rebased.
Thanks.

File: src/kv/MemDB.cc [hunk 6]
```diff
@@ -397,7 +401,7 @@ void MemDB::MDBWholeSpaceIteratorImpl::fill_current()
 
 bool MemDB::MDBWholeSpaceIteratorImpl::valid()
 {
-  if (m_key_value.first.empty()) {
+  if (!(*is_valid) || m_key_value.first.empty()) {
```
Comment (chhabaramesh) [inline src/kv/MemDB.cc]: Rest looks fine, But I am not sure if current iterator becomes invalid then we should return invalid or restart iterator from current position. And if we return invalid, will bluestore take care of this and restart new iterator. I will confirm soon,

Comment (chhabaramesh) [inline src/kv/MemDB.cc]: I did not notice @liewegas  comment . His comment supersedes mine.

Comment (chhabaramesh) [inline src/kv/MemDB.cc]: After keeping track of seqno, I think just restarting the iterator with m_key_value.first  as key should work. m_iter = m_btree_p->lower_bound(k);

Comment (tanghaodong25) [inline src/kv/MemDB.cc]: @chhabaramesh @liewegas Thanks for your comments.

I'm not sure if bluestore will retry to get a valid iterator before, so I just add a bool. Then I will fix it with seqno:)

File: src/kv/MemDB.h [hunk 7]
```diff
@@ -132,12 +133,16 @@ class MemDB : public KeyValueDB
...
       std::mutex *m_btree_lock_p;
+      uint64_t *btree_seq;
+      uint64_t current_seq;
 
   public:
     MDBWholeSpaceIteratorImpl(btree::btree_map<std::string, bufferptr> *btree_p,
-                             std::mutex *btree_lock_p) {
+                             std::mutex *btree_lock_p, uint64_t *btree_seq_) {
         m_btree_p = btree_p;
         m_btree_lock_p = btree_lock_p;
```
Comment (chhabaramesh) [inline src/kv/MemDB.h]: I think we need to take btree lock here.

File: src/kv/MemDB.h [hunk 8]
```diff
@@ -147,12 +152,13 @@ class MemDB : public KeyValueDB
     int seek_to_first(const std::string &k);
     int seek_to_last(const std::string &k);
 
-    int seek_to_first() { return seek_to_first(NULL); };
-    int seek_to_last() { return seek_to_last(NULL); };
+    int seek_to_first() { return seek_to_first(""); };
```
Comment (chhabaramesh) [inline src/kv/MemDB.h]: NULL is not good for cpp that is true but can we make empty string more explicit like std::string("NULL") ?

File: src/kv/MemDB.h [hunk 9]
```diff
@@ -132,12 +133,16 @@ class MemDB : public KeyValueDB
       std::pair<string, bufferlist> m_key_value;
       btree::btree_map<std::string, bufferptr> *m_btree_p;
       std::mutex *m_btree_lock_p;
+      uint64_t *btree_seq;
+      uint64_t current_seq;
```
Comment (chhabaramesh) [inline src/kv/MemDB.h]: Rename it to something : global_seq_no and my_seq_no or this_seq_no. Current seq_node is confusing .

**Comment by chhabaramesh** [comment]
@tanghaodong25  I have some minor comments on latest code. Please fix those and retest.  Overall looks fine to me.

**Comment by tanghaodong25** [comment]
@chhabaramesh  Thanks. :)

File: src/kv/MemDB.cc [hunk 10]
```diff
@@ -472,24 +494,24 @@ int MemDB::MDBWholeSpaceIteratorImpl::seek_to_first(const std::string &k)
 {
   std::lock_guard<std::mutex> l(*m_btree_lock_p);
   free_last();
-  if (k.empty()) {
+  if (k == "NULL") {
```
Comment (liewegas) [inline src/kv/MemDB.cc]: pretty sure k.empty() is the right thing here...

File: src/kv/MemDB.cc [hunk 11]
```diff
@@ -472,24 +494,24 @@ int MemDB::MDBWholeSpaceIteratorImpl::seek_to_first(const std::string &k)
...
   return 0;
 }
 
 int MemDB::MDBWholeSpaceIteratorImpl::seek_to_last(const std::string &k)
 {
   std::lock_guard<std::mutex> l(*m_btree_lock_p);
-
   free_last();
-  if (k.empty()) {
+  if (k == "NULL") {
```
Comment (liewegas) [inline src/kv/MemDB.cc]: and here

File: src/kv/MemDB.h [hunk 12]
```diff
@@ -147,12 +153,13 @@ class MemDB : public KeyValueDB
     int seek_to_first(const std::string &k);
     int seek_to_last(const std::string &k);
 
-    int seek_to_first() { return seek_to_first(NULL); };
-    int seek_to_last() { return seek_to_last(NULL); };
+    int seek_to_first() { return seek_to_first("NULL"); };
+    int seek_to_last() { return seek_to_last("NULL"); };
```
Comment (liewegas) [inline src/kv/MemDB.h]: seek_to_first(string()); etc

Comment (liewegas) [inline src/kv/MemDB.h]: actaully, why change this at all?  doesn't seem related to the fix.

File: src/kv/MemDB.cc [hunk 13]
```diff
@@ -478,11 +478,12 @@ int MemDB::MDBWholeSpaceIteratorImpl:: prev()
...
   }
   free_last();
-  m_iter--;
-  if (m_iter != m_btree_p->end()) {
+  if (m_iter != m_btree_p->begin()) {
+    m_iter--;
     fill_current();
     return 0;
   } else {
+    m_iter--;
```
Comment (liewegas) [inline src/kv/MemDB.cc]: if m_iter == m_btree_p->begin() we shouldn't --, right?  just return -1?

**Comment by tanghaodong25** [comment]
@liewegas  updated. Thanks.

File: src/kv/MemDB.cc [hunk 14]
```diff
@@ -477,17 +499,17 @@ int MemDB::MDBWholeSpaceIteratorImpl::seek_to_first(const std::string &k)
   } else {
     m_iter = m_btree_p->lower_bound(k);
   }
-
+  
   if (m_iter == m_btree_p->end()) {
     return -1;
   }
+  fill_current();
```
Comment (liewegas) [inline src/kv/MemDB.cc]: these two changes are fixing an unrelated bug in seek_to_*, right?  Can you put them in a separate commit?

**Comment by liewegas** [comment]
It would be great to have a test for this in test/objectstore/test_kv.cc... :)

**Comment by chhabaramesh** [comment]
@tanghaodong25, please also run ceph_test_objectstore.

**Comment by tanghaodong25** [comment]
@chhabaramesh  @liewegas  All test in "ceph_test_objectstore" and "ceph_test_keyvaluedb" passed. :)
